### PR TITLE
Set no-trailing-spaces to ignore entirely whitespace lines

### PR DESCRIPTION
**Pull request by andrewvaughan**
Closes issue #2303

### CODE REVIEW

**Comment by nzakas** [comment]
Thanks. Please read over our contribution guidelines for how to format your pull request: http://eslint.org/docs/developer-guide/contributing.html

**Comment by andrewvaughan** [comment]
The PR has been reformatted - please review.

File: lib/rules/no-trailing-spaces.js [hunk 1]
```diff
@@ -10,7 +10,12 @@
 
 module.exports = function(context) {
 
-    var TRAILER = "[ \t\u00a0\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u200b\u2028\u2029\u3000]+$";
+    var BLANK_PREFIX = "[^( |\t)]",
+        TRAILER = "[ \t\u00a0\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u200b\u2028\u2029\u3000]+$";
+
+    var options = context.options[0] || {},
+        skipBlankLines = +!!options.skipBlankLines; // Defaults to false
```
Comment (nzakas) [inline lib/rules/no-trailing-spaces.js]: The `+` is unnecessary here, please remove.

File: tests/lib/rules/no-trailing-spaces.js [hunk 2]
```diff
@@ -19,8 +19,42 @@ var eslintTester = new ESLintTester(eslint);
 eslintTester.addRuleTest("lib/rules/no-trailing-spaces", {
 
     valid: [
-        "var a = 5;",
-        "var a = 5,\n    b = 3;"
+        {
+            code: "var a = 5;",
+            options: [{}]
+        },
```
Comment (nzakas) [inline tests/lib/rules/no-trailing-spaces.js]: Include a couple tests that do not specify options.

Comment (andrewvaughan) [inline tests/lib/rules/no-trailing-spaces.js]: There are - can you please explain what's missing further?

File: lib/rules/no-trailing-spaces.js (hunk 1 above)
Comment (andrewvaughan) [inline lib/rules/no-trailing-spaces.js]: Are you referring to TRAILER?  That was the original code.  I can remove if needed, though.

File: tests/lib/rules/no-trailing-spaces.js (hunk 2 above)
Comment (nzakas) [inline tests/lib/rules/no-trailing-spaces.js]: What I mean is omit `options` altogether. There's a difference between passing an array with an empty object and not passing anything.

Comment (andrewvaughan) [inline tests/lib/rules/no-trailing-spaces.js]: Sure - I would review some of the other tests (specifically key-spacing) as I used that for reference when building out the code.  I will touch on these tonight.

**Comment by andrewvaughan** [comment]
I've made appropriate changes - please review again when convenient.

File: tests/lib/rules/no-trailing-spaces.js (hunk 2 above)
Comment (nzakas) [inline tests/lib/rules/no-trailing-spaces.js]: I'm still not seeing any tests that don't have `options` defined.

Comment (andrewvaughan) [inline tests/lib/rules/no-trailing-spaces.js]: Lines 30-35 and 91-104 in tests/lib/rules/no-trailing-spaces.js

I'm afraid I can't spend any more time on this issue - please take over as you see fit.

File: tests/lib/rules/no-trailing-spaces.js [hunk 3]
```diff
@@ -19,8 +19,48 @@ var eslintTester = new ESLintTester(eslint);
...
+        {
```
Comment (nzakas) [inline tests/lib/rules/no-trailing-spaces.js]: Ah, somehow completely missed that.

**Comment by nzakas** [comment]
This looks good. We also need the docs to be updated with the new option. Is this something you have time to do?

**Comment by nzakas** [comment]
Merged manually with the updated documentation

**Pull request closed without merging**